src/ConstantBounds.cpp

```diff
@@ -0,0 +1,1141 @@
+        }
+
+        // No luck, return the mutated Max
+        if (!a_new.same_as(op->a) || !b_new.same_as(op->b)) {
+            return Max::make(std::move(a_new), std::move(b_new));
+        } else {
+            return op;
+        }
+    }
+
+    Expr visit(const Select *op) override {
+        // Do not visit condition, don't care about those variables.
+        Expr true_value = mutate(op->true_value);
+        Expr false_value = mutate(op->false_value);
+        if (true_value.same_as(op->true_value) &&
+            false_value.same_as(op->false_value)) {
+            return op;
+        }
+        return Select::make(op->condition, std::move(true_value), std::move(false_value));
+    }
+
+    // Do nothing for Loads, Ramps, Broadcasts, Calls, Shuffles, VectorReduce
+    Expr visit(const Load *op) override {
+        return op;
+    }
+
+    Expr visit(const Ramp *op) override {
+        return op;
+    }
+
+    Expr visit(const Broadcast *op) override {
+        return op;
+    }
+
+    Expr visit(const Call *op) override {
+        return op;
+    }
+
+    Expr visit(const Shuffle *op) override {
+        return op;
+    }
+
+    Expr visit(const VectorReduce *op) override {
+        return op;
+    }
+
+public:
+    StripUnboundedTerms(Direction _direction, const Scope<Interval> *_scope_ptr,
+                        const std::map<std::string, int> &_var_uses)
+        : direction(_direction), scope_ptr(_scope_ptr), var_uses(_var_uses) {
+    }
+};
+
+/*
+ * Visitor for reordering summations and removing like-terms from min/max/select nodes.
+ * This is an exact method, no approximations.
+ * Summations are represented as a vector of affine terms.
+ */
+class ReorderTerms : public IRGraphMutator {
+    using IRGraphMutator::visit;
+
+    // Directly taken from Simplify_Internal.h. May want to make this a generic function.
+    HALIDE_ALWAYS_INLINE
+    bool should_commute(const Expr &a, const Expr &b) {
+        if (a.node_type() < b.node_type()) {
+            return true;
+        }
+        if (a.node_type() > b.node_type()) {
+            return false;
+        }
+
+        if (a.node_type() == IRNodeType::Variable) {
+            const Variable *va = a.as<Variable>();
+            const Variable *vb = b.as<Variable>();
+            return va->name.compare(vb->name) > 0;
+        }
+
+        return false;
+    }
+
+    // This is very similar to code in LICM, but we don't care about depth.
+    struct AffineTerm {
+        Expr expr;
+        int coefficient;
+    };
+
+    // Simple helper function for negating a term.
+    AffineTerm negate(const AffineTerm &term) {
+        return AffineTerm{term.expr, -term.coefficient};
+    }
+
+    // Useful for extracting an affine term from an (x * c) term
+    bool is_mul_by_int_const(const Expr &expr, AffineTerm &term) {
+        const Mul *mul = expr.as<Mul>();
+        if (mul) {
+            if (const IntImm *a = mul->a.as<IntImm>()) {
+                term.coefficient = a->value;
```

Comment:
This should only match is a->value and -(a->value) are representable as int, because you negate it above.